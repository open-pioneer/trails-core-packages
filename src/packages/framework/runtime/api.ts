// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0

/* eslint-disable @typescript-eslint/no-explicit-any */
export type ApiMethod = (...args: any[]) => any;

/**
 * A record of exposed functions.
 */
export type ApiMethods = Record<string, ApiMethod>;

/**
 * Provides a set of methods to the application's web component API.
 */
export interface ApiExtension {
    /**
     * Returns a set of methods that will be added to the web component's API.
     */
    getApiMethods(): Promise<ApiMethods>;
}

/**
 * A service provided by the system.
 * Used by the runtime to assemble the public facing API.
 */
export interface ApiService {
    /**
     * Called by the runtime to gather methods that should be available from the web component's API.
     */
    getApi(): Promise<ApiMethods>;
}

/**
 * A service provided by the system, useful for accessing values that are global to the application.
 */
export interface ApplicationContext {
    /**
     * The web component's host element.
     * This dom node can be accessed by the host site.
     */
    getHostElement(): HTMLElement;

    /**
     * The current web component's shadow root.
     */
    getShadowRoot(): ShadowRoot;

    /**
     * The node containing the rest of the application _inside_ the current web component's shadow dom.
     */
    getApplicationContainer(): HTMLElement;

    /**
     * Returns the current locale of the application.
     *
     * E.g. "de-DE"
     */
    getLocale(): string;

    /**
     * Returns the locales supported by the application.
     *
     * E.g. ["de", "en"]
     */
    getSupportedLocales(): readonly string[];
}

declare module "./ServiceRegistry" {
    interface ServiceRegistry {
        "runtime.ApiService": ApiService;
        "runtime.ApplicationContext": ApplicationContext;
    }
}
