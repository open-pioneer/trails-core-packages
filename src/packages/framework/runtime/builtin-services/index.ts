import { createEmptyI18n } from "../i18n";
import { PackageRepr } from "../service-layer/PackageRepr";
import {
    createConstructorFactory,
    createFunctionFactory,
    ServiceRepr
} from "../service-layer/ServiceRepr";
import { ApiServiceImpl } from "./ApiServiceImpl";
import { ApplicationContextImpl, ApplicationContextProperties } from "./ApplicationContextImpl";

export const RUNTIME_PACKAGE_NAME = "@open-pioneer/runtime";
export const RUNTIME_API_EXTENSION = "integration.ApiExtension";
export const RUNTIME_API_SERVICE = "runtime.ApiService";
export const RUNTIME_APPLICATION_CONTEXT = "runtime.ApplicationContext";

export type BuiltinPackageProperties = ApplicationContextProperties;

/**
 * Creates the builtin package containing the builtin services.
 *
 * The runtime package should not use services in the `build.config.mjs`
 * for a clean bootstrapping procedure (it will be instantiated without
 * generated application metadata in tests).
 *
 * This function is called as part of the service layer startup.
 * The package produced here is always part of the application.
 */
export function createBuiltinPackage(properties: BuiltinPackageProperties): PackageRepr {
    const i18n = createEmptyI18n();
    const apiService = new ServiceRepr({
        name: "ApiServiceImpl",
        packageName: RUNTIME_PACKAGE_NAME,
        factory: createConstructorFactory(ApiServiceImpl),
        intl: i18n,
        interfaces: [
            {
                interfaceName: RUNTIME_API_SERVICE,
                qualifier: "builtin"
            }
        ],
        dependencies: [
            {
                referenceName: "providers",
                interfaceName: RUNTIME_API_EXTENSION,
                all: true
            }
        ]
    });
    const appContext = new ServiceRepr({
        name: "ApplicationContextImpl",
        packageName: RUNTIME_PACKAGE_NAME,
        factory: createFunctionFactory(
            (options) => new ApplicationContextImpl(options, properties)
        ),
        intl: i18n,
        interfaces: [
            {
                interfaceName: RUNTIME_APPLICATION_CONTEXT,
                qualifier: "builtin"
            }
        ]
    });

    return new PackageRepr({
        name: RUNTIME_PACKAGE_NAME,
        services: [apiService, appContext],
        intl: i18n
    });
}
