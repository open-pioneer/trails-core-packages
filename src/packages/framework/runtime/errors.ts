export enum ErrorId {
    // Problems with application metadata
    INVALID_METADATA = "runtime:invalid-metadata",
    INVALID_PROPERTY_NAME = "runtime:invalid-property-name",
    REQUIRED_PROPERTY = "runtime:required-property",

    // Application lifecycle
    NOT_MOUNTED = "runtime:element-not-mounted",
    UNSUPPORTED_LOCALE = "runtime:unsupported-locale",
    CONFIG_RESOLUTION_FAILED = "runtime:config-resolution-failed",

    // Service layer
    INTERFACE_NOT_FOUND = "runtime:interface-not-found",
    AMBIGUOUS_DEPENDENCY = "runtime:ambiguous-dependency",
    UNDECLARED_DEPENDENCY = "runtime:undeclared-dependency",
    SERVICE_CONSTRUCTION_FAILED = "runtime:service-construction-failed",
    SERVICE_DESTRUCTION_FAILED = "runtime:service-destruction-failed",
    DUPLICATE_INTERFACE = "runtime:duplicate-interface",
    DEPENDENCY_CYCLE = "runtime:dependency-cycle",

    // Web component API
    DUPLICATE_API_METHODS = "runtime:duplicate-api-methods",

    // Internal
    INTERNAL = "runtime:internal-error"
}
