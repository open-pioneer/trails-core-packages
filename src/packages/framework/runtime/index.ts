// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
export * from "./api";
export { type PackageIntl } from "./i18n";
export {
    type ApplicationElement,
    type ApplicationElementConstructor,
    type ApplicationProperties,
    type ApplicationConfig,
    type CustomElementOptions,
    type ConfigContext,
    type RawApplicationProperties,
    createCustomElement
} from "./CustomElement";
export * from "./Service";
export * from "./ServiceRegistry";
export * from "./PropertiesRegistry";

export { ApiServiceImpl } from "./builtin-services/ApiServiceImpl";
