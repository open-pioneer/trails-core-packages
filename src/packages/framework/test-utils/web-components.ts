import { waitFor, within } from "@testing-library/dom";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const GLOBAL = global as any;

function nextId(): number {
    GLOBAL["__WEB_COMPONENT_COUNT__"] ??= 0;
    const id = (GLOBAL["__WEB_COMPONENT_COUNT__"] += 1);
    return id;
}

/**
 * Renders the given component into the DOM and returns the new node.
 */
export async function renderComponent(
    component: CustomElementConstructor | string,
    options?: {
        container?: HTMLElement;
        attributes?: Record<string, string>;
    }
) {
    const tag = typeof component === "string" ? component : defineComponent(component);
    const container = options?.container ?? document.body;
    const node = container.ownerDocument.createElement(tag);
    for (const [key, value] of Object.entries(options?.attributes ?? {})) {
        node.setAttribute(key, value);
    }
    container.innerHTML = "";
    container.appendChild(node);
    return {
        node
    };
}

/**
 * Renders the given component into the DOM (via {@link renderComponent}) and works on its shadow root.
 * In order to use this function, the shadow root should be attached as `open` from inside the component class.
 *
 * After the shadow root has been accessed, the function searches (and waits) for an inner root container (by default `.pioneer-root`)
 * where all other searches should be executed.
 *
 * Returns the shadow root, the inner container, and a bound queries object that automatically searches in the inner container
 * instead of the whole document.
 */
export async function renderComponentShadowDOM(
    component: CustomElementConstructor | string,
    options?: {
        container?: HTMLElement;
        attributes?: Record<string, string>;
        innerContainerSelector?: string;
    }
) {
    const { node } = await renderComponent(component, options);
    const selector = options?.innerContainerSelector ?? ".pioneer-root";
    const { shadowRoot, innerContainer } = await waitFor(() => {
        const shadowRoot = node.shadowRoot;
        if (!shadowRoot) {
            throw new Error("Shadow root not attached.");
        }
        const innerContainer = shadowRoot.querySelector<HTMLElement>(selector);
        if (!innerContainer) {
            throw new Error("Inner container not ready.");
        }
        return { shadowRoot, innerContainer };
    });

    const queries = within(innerContainer);
    return {
        node,
        shadowRoot,
        innerContainer,
        queries
    };
}

/**
 * Defines the given component and returns the unique tag name.
 * This helper is necessary because tag names must not collide in the global elements registry (and
 * un-registration is not possible).
 */
export function defineComponent(
    component: CustomElementConstructor,
    options?: {
        nameHint?: string;
    }
) {
    const nameHint = options?.nameHint ?? "test-component";
    const tag = `${nameHint}-${nextId()}`;
    customElements.define(tag, component);
    return tag;
}
