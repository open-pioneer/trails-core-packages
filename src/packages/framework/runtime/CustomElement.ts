// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { ComponentType } from "react";
import {
    createAbortError,
    createLogger,
    createManualPromise,
    destroyResource,
    Error,
    isAbortError,
    ManualPromise,
    Resource,
    throwAbortError
} from "@open-pioneer/core";
import { ErrorId } from "./errors";
import { ApplicationMetadata, ObservableBox, PackageMetadata } from "./metadata";
import { PackageRepr, createPackages } from "./service-layer/PackageRepr";
import { ServiceLayer } from "./service-layer/ServiceLayer";
import { getErrorChain } from "@open-pioneer/core";
import { ReactIntegration } from "./react-integration/ReactIntegration";
import { ApiMethods, ApiService } from "./api";
import { createBuiltinPackage, RUNTIME_API_SERVICE } from "./builtin-services";
import { ReferenceSpec } from "./service-layer/InterfaceSpec";
import { PropertiesRegistry } from "./PropertiesRegistry";
import { AppI18n, initI18n } from "./i18n";
const LOG = createLogger("runtime:CustomElement");

/**
 * Options for the {@link createCustomElement} function.
 */
export interface CustomElementOptions {
    /**
     * Rendered UI component.
     */
    component?: ComponentType<Record<string, string>>;

    /**
     * Application metadata (packages, services etc.).
     * This is usually autogenerated by importing the virtual `"open-pioneer:app"` module.
     */
    appMetadata?: ApplicationMetadata;

    /**
     * Application defined configuration.
     * All instances of the web component will share this static configuration.
     *
     * These will override default properties of the application's packages.
     */
    config?: ApplicationConfig;

    /**
     * Function to provide additional application defined configuration parameters.
     *
     * Compared to {@link config}, this function receives a context object
     * that allows the developer to provide dynamic properties on a per-application basis.
     *
     * Parameters returned by this function take precedence over the ones defined by {@link config}.
     */
    resolveConfig?(ctx: ConfigContext): Promise<ApplicationConfig | undefined>;

    /**
     * Whether the shadow root element is accessible from the outside.
     * Defaults to `false` in production mode and `true` during development to make testing easier.
     */
    openShadowRoot?: boolean;
}

/**
 * A context object that is passed to the `resolveProperties` function.
 */
export interface ConfigContext {
    /**
     * Returns an attribute from the application's root node.
     */
    getAttribute(name: string): string | undefined;
}

/**
 * Runtime application configuration.
 */
export interface ApplicationConfig {
    /**
     * Set this value to a locale string (e.g. "en") to for the application's locale.
     * The default behavior is to choose an appropriate locale for the current user based
     * on the browser's settings.
     *
     * The locale must be supported by the application.
     */
    locale?: string | undefined;

    /**
     * Properties specified here will override default properties of the application's packages.
     */
    properties?: ApplicationProperties | undefined;
}

/**
 * Allows the application to override default properties in all packages.
 */
export interface RawApplicationProperties {
    /**
     * Key: the name of the package.
     * Value: A record of configuration properties (key/value pairs).
     *
     * Properties will override default property values in the package.
     */
    [packageName: string]: Record<string, unknown>;
}

/**
 * Allows the application to override default properties in all packages.
 *
 * Properties are typed when the package contains type definitions for them.
 */
export type ApplicationProperties = RawApplicationProperties & Partial<PropertiesRegistry>;

/**
 * The interface implemented by web components produced via {@link createCustomElement}.
 */
export interface ApplicationElement extends HTMLElement {
    /** Resolves to the element's API when the application has started. */
    when(): Promise<ApiMethods>;
}

/**
 * The class returned by a call to {@link createCustomElement}.
 */
export interface ApplicationElementConstructor {
    new (): ApplicationElement;
}

/**
 * Creates a new custom element class (web component) that can be registered within a DOM.
 *
 * @example
 * ```ts
 * import * as appMetadata from "open-pioneer:app";
 *
 * const CustomElementClazz = createCustomElement({
 *   component: <div>Hello World!</div>,
 *   appMetadata
 * });
 * customElements.define("sample-element", CustomElementClazz);
 * ```
 */
export function createCustomElement(options: CustomElementOptions): ApplicationElementConstructor {
    class PioneerApplication extends HTMLElement implements ApplicationElement {
        #shadowRoot: ShadowRoot;
        #state: ElementState | undefined;

        static get observedAttributes(): string[] {
            return [];
        }

        constructor() {
            super();

            const mode = options.openShadowRoot ?? import.meta.env.DEV ? "open" : "closed";
            this.#shadowRoot = this.attachShadow({
                mode: mode
            });
        }

        connectedCallback() {
            LOG.debug("Launching application");

            if (this.#state) {
                this.#state.destroy();
            }

            this.#state = new ElementState(this, this.#shadowRoot, options);
            this.#state.start();
        }

        disconnectedCallback() {
            LOG.debug("Shutting down application");

            this.#state?.destroy();
            this.#state = undefined;

            LOG.debug("Application destroyed");
        }

        when() {
            if (!this.#state) {
                return Promise.reject(
                    new Error(
                        ErrorId.NOT_MOUNTED,
                        "Cannot use the application's API because the HTML element has not yet been mounted into the DOM."
                    )
                );
            }

            return this.#state.whenAPI();
        }
    }
    return PioneerApplication;
}

class ElementState {
    private hostElement: HTMLElement;
    private shadowRoot: ShadowRoot;
    private options: CustomElementOptions;

    // Public API
    private apiPromise: ManualPromise<ApiMethods> | undefined; // Present when callers are waiting for the API
    private api: ApiMethods | undefined; // Present once started

    private state = "not-started" as "not-started" | "starting" | "started" | "destroyed";
    private locale: string | undefined;
    private container: HTMLDivElement | undefined;
    private config: ApplicationConfig | undefined;
    private serviceLayer: ServiceLayer | undefined;
    private reactIntegration: ReactIntegration | undefined;
    private stylesWatch: Resource | undefined;

    constructor(hostElement: HTMLElement, shadowRoot: ShadowRoot, options: CustomElementOptions) {
        this.hostElement = hostElement;
        this.shadowRoot = shadowRoot;
        this.options = options;
    }

    start() {
        if (this.state !== "not-started") {
            throw new Error(ErrorId.INTERNAL, `Cannot start element in state '${this.state}'`);
        }

        this.state = "starting";
        this.startImpl().catch((e) => {
            // TODO: Error splash?
            this.destroy();
            if (!isAbortError(e)) {
                logError(e);
            }
        });
    }

    destroy() {
        this.state = "destroyed";
        this.apiPromise?.reject(createAbortError());
        this.reactIntegration = destroyResource(this.reactIntegration);
        this.shadowRoot.replaceChildren();
        this.container = undefined;
        this.serviceLayer = destroyResource(this.serviceLayer);
        this.stylesWatch = destroyResource(this.stylesWatch);
    }

    whenAPI(): Promise<ApiMethods> {
        if (this.api) {
            return Promise.resolve(this.api);
        }

        const apiPromise = (this.apiPromise ??= createManualPromise());
        return apiPromise.promise;
    }

    private async startImpl() {
        const { options, shadowRoot, hostElement } = this;

        // Resolve custom application config
        const config = (this.config = await gatherConfig(hostElement, options));
        this.checkAbort();
        LOG.debug("Application config is", config);

        // Decide on locale and load i18n messages (if any).
        const i18n = await initI18n(options.appMetadata, config.locale);
        this.checkAbort();

        // Setup application root node in the shadow dom
        const container = (this.container = createContainer(i18n.locale));
        const styles = this.initStyles();
        shadowRoot.replaceChildren(container, styles);

        // Launch the service layer
        const { serviceLayer, packages } = this.initServiceLayer({
            container,
            properties: config.properties,
            i18n
        });
        await this.initAPI(serviceLayer);
        this.checkAbort();

        // Launch react
        this.reactIntegration = new ReactIntegration({
            rootNode: container,
            container: shadowRoot,
            serviceLayer,
            packages
        });
        this.render();
        this.state = "started";

        LOG.debug("Application started");
    }

    private render() {
        this.reactIntegration?.render(this.options.component ?? emptyComponent, {});
    }

    private initStyles() {
        const styles = this.options.appMetadata?.styles;
        const styleNode = document.createElement("style");
        applyStyles(styleNode, styles);
        if (import.meta.hot) {
            this.stylesWatch = styles?.on?.("changed", () => {
                LOG.debug("Application styles changed");
                applyStyles(styleNode, styles);
            });
        }
        return styleNode;
    }

    private initServiceLayer(config: {
        container: HTMLDivElement;
        properties: ApplicationProperties;
        i18n: AppI18n;
    }) {
        const options = this.options;
        const { container, properties, i18n } = config;
        const packageMetadata = options.appMetadata?.packages ?? {};
        const builtinPackage = createBuiltinPackage({
            host: this.hostElement,
            shadowRoot: this.shadowRoot,
            container: container,
            locale: i18n.locale,
            supportedLocales: i18n.supportedLocales
        });
        const { serviceLayer, packages } = createServiceLayer({
            packageMetadata,
            builtinPackage,
            properties,
            i18n
        });
        this.serviceLayer = serviceLayer;

        if (LOG.isDebug()) {
            LOG.debug("Launching service layer with packages", Object.fromEntries(packages));
        }
        serviceLayer.start();
        return { serviceLayer, packages };
    }

    private async initAPI(serviceLayer: ServiceLayer) {
        const result = serviceLayer.getService(
            "@open-pioneer/runtime",
            {
                interfaceName: RUNTIME_API_SERVICE
            },
            { ignoreDeclarationCheck: true }
        );
        if (result.type !== "found") {
            throw new Error(
                ErrorId.INTERNAL,
                `Failed to find instance of 'runtime.ApiService' (result type '${result.type}').` +
                    ` This is a builtin service that must be present exactly once.`
            );
        }

        const apiService = result.value.getInstanceOrThrow() as ApiService;
        try {
            const api = (this.api = await apiService.getApi());
            LOG.debug("Application API initialized to", api);
            this.apiPromise?.resolve(api);
        } catch (e) {
            throw new Error(ErrorId.INTERNAL, "Failed to gather the application's API methods.", {
                cause: e
            });
        }
    }

    private checkAbort() {
        if (this.state === "destroyed") {
            throwAbortError();
        }
    }
}

function createContainer(locale: string) {
    // Setup application root node in the shadow dom
    const container = document.createElement("div");
    container.classList.add("pioneer-root");
    container.style.minHeight = "100%";
    container.style.height = "100%";
    if (locale) {
        container.lang = locale;
    }
    return container;
}

function createServiceLayer(config: {
    packageMetadata: Record<string, PackageMetadata> | undefined;
    properties: ApplicationProperties;
    builtinPackage: PackageRepr;
    i18n: AppI18n;
}) {
    const { packageMetadata, properties, builtinPackage, i18n } = config;

    let packages: PackageRepr[];
    try {
        packages = createPackages(packageMetadata ?? {}, i18n, properties);
    } catch (e) {
        throw new Error(ErrorId.INVALID_METADATA, "Failed to parse package metadata.", {
            cause: e
        });
    }

    // Add builtin services defined within this package.
    {
        const index = packages.findIndex((pkg) => pkg.name === builtinPackage.name);
        if (index >= 0) {
            packages.splice(index, 1);
        }
        packages.push(builtinPackage);
    }

    const forcedReferences: ReferenceSpec[] = [
        {
            interfaceName: RUNTIME_API_SERVICE
        }
    ];
    const serviceLayer = new ServiceLayer(packages, forcedReferences);
    return {
        packages: new Map(packages.map((pkg) => [pkg.name, pkg])),
        serviceLayer: serviceLayer
    };
}

/**
 * Gathers application properties by reading them from the options object
 * and by (optionally) invoking the `resolveProperties` hook.
 */
async function gatherConfig(hostElement: HTMLElement, options: CustomElementOptions) {
    let configs: ApplicationConfig[];
    try {
        configs = [
            options.config ?? {},
            (await options.resolveConfig?.({
                getAttribute(name) {
                    return hostElement.getAttribute(name) ?? undefined;
                }
            })) ?? {}
        ];
    } catch (e) {
        throw new Error(
            ErrorId.CONFIG_RESOLUTION_FAILED,
            "Failed to resolve application properties.",
            {
                cause: e
            }
        );
    }

    return mergeConfigs(configs);
}

/**
 * Merges application configurations into a single object.
 * Properties / config parameters at a later position overwrite properties from earlier ones.
 */
function mergeConfigs(configs: ApplicationConfig[]): Required<ApplicationConfig> {
    // Merge simple values by assigning them in order
    const mergedConfig: Required<ApplicationConfig> = Object.assign(
        {
            locale: undefined,
            properties: {}
        } satisfies ApplicationConfig,
        ...configs
    );

    // Deep merge for application properties
    const mergedProperties: ApplicationProperties = (mergedConfig.properties = {});
    for (const config of configs) {
        for (const [packageName, packageProperties] of Object.entries(config.properties ?? {})) {
            const mergedPackageProps = (mergedProperties[packageName] ??= {});
            Object.assign(mergedPackageProps, packageProperties);
        }
    }

    return mergedConfig;
}

const DISABLE_INHERIT = ":host { all: initial; display: block; }";

// Applies application styles to the given style node.
// Can be called multiple times in development mode to implement hot reloading.
function applyStyles(styleNode: HTMLStyleElement, styles: ObservableBox<string> | undefined) {
    const cssValue = styles?.value ?? "";
    const cssNode = document.createTextNode([DISABLE_INHERIT, cssValue].join("\n"));
    styleNode.replaceChildren(cssNode);
}

function logError(e: unknown) {
    if (e instanceof Error) {
        const chain = getErrorChain(e).reverse();
        if (chain.length === 1) {
            console.error(e);
            return;
        }

        let n = 1;
        for (const error of chain) {
            console.error(`#${n}`, error);
            ++n;
        }
    } else {
        console.error("Unexpected error", e);
    }
}

function emptyComponent() {
    return null;
}
