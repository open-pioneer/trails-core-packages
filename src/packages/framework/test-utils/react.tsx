import { createIntl, createIntlCache, IntlShape } from "@formatjs/intl";
import { Service } from "@open-pioneer/runtime";
import {
    PackageContext as InternalPackageContext,
    PackageContextMethods
} from "@open-pioneer/runtime/react-integration";
import { FC, ReactNode, useMemo } from "react";

export type AnyService = Service<Record<string, unknown>>;

export interface PackageContextProviderProps {
    /** Interface implementations, keyed by interface name. */
    services?: {
        [interfaceName: string]: AnyService;
    };

    /** Interface implementations, keyed by interface name and then by qualifier. */
    qualifiedServices?: {
        [interfaceName: string]: {
            [qualifier: string]: AnyService;
        };
    };

    /** Package properties (keyed by package name). */
    properties?: {
        [packageName: string]: Record<string, unknown>;
    };

    /**
     * The locale for i18n messages and formatting.
     *
     * @default "en"
     */
    locale?: string;

    /**
     * I18n messages for packages
     */
    messages?: {
        /**
         * I18n messages as (messageId, message) entries.
         *
         * @default {}
         */
        [packageName: string]: Record<string, string>;
    };

    /** Children to render */
    children?: ReactNode;
}

/**
 * Injects services and properties into the component tree.
 * React components using pioneer hooks like `useService` and `useProperties`
 * will receive the mocked properties here instead.
 */
export const PackageContextProvider: FC<PackageContextProviderProps> = (props) => {
    const { children, ...rest } = props;
    const contextMethods = useMemo(() => createPackageContextMethods(rest), [rest]);
    return (
        <InternalPackageContext.Provider value={contextMethods}>
            {children}
        </InternalPackageContext.Provider>
    );
};

function createPackageContextMethods(
    options: Omit<PackageContextProviderProps, "children">
): PackageContextMethods {
    const services = options?.services ?? {};
    const qualifiedServices = options?.qualifiedServices ?? {};
    const properties = options?.properties ?? {};
    const locale = options?.locale ?? "en";
    const messages = options?.messages ?? {};
    const cachedIntl: Record<string, IntlShape> = {};
    return {
        getService(packageName, interfaceName, options) {
            if (!options.qualifier) {
                const service = services[interfaceName];
                if (service) {
                    return service;
                }
                const qualified = qualifiedServices[interfaceName];
                if (qualified) {
                    for (const value of Object.values(qualified)) {
                        return value; // Return any implementation
                    }
                }
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}'. Update the configuration of PackageContextProvider.`
                );
            }

            const qualifiedService = qualifiedServices[interfaceName]?.[options.qualifier];
            if (!qualifiedService) {
                throw new Error(
                    `Interface name not bound for testing: '${interfaceName}' (qualifier '${options.qualifier}'). Update the configuration of PackageContextProvider.`
                );
            }
            return qualifiedService;
        },
        getServices(packageName, interfaceName) {
            const unqualified = services[interfaceName];
            const results = Object.values(qualifiedServices[interfaceName] ?? {});
            if (unqualified) {
                results.push(unqualified);
            }
            return results;
        },
        getProperties(packageName) {
            const packageProperties = properties[packageName];
            if (!packageProperties) {
                throw new Error(
                    `No properties for package '${packageName}' bound for testing. Update the configuration of PackageContextProvider.`
                );
            }
            return packageProperties;
        },
        getIntl(packageName) {
            const initIntl = () => {
                const packageMessages = messages[packageName];
                const cache = createIntlCache();
                return createIntl(
                    {
                        locale,
                        messages: packageMessages
                    },
                    cache
                );
            };
            return (cachedIntl[packageName] ??= initIntl());
        }
    };
}
