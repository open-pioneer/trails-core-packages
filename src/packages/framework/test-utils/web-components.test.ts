// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
/**
 * @vitest-environment jsdom
 */
import { createCustomElement } from "@open-pioneer/runtime";
import { findByTestId } from "@testing-library/dom";
import { createElement } from "react";
import { afterEach, expect, it } from "vitest";
import { defineComponent, renderComponent, renderComponentShadowDOM } from "./web-components";

afterEach(() => {
    document.body.innerHTML = "";
});

it("should render a web component into the dom", async function () {
    class CustomComponent extends HTMLElement {
        constructor() {
            super();
        }

        connectedCallback() {
            this.innerHTML = `<span data-testid="foo" class="hello">Hello World</span>`;
        }
    }

    const { node } = await renderComponent(CustomComponent);
    const span = await findByTestId(node, "foo");
    expect(span.className).toEqual("hello");
    expect(span.textContent).toEqual("Hello World");
});

it("should render the same component multiple times (needs shared tag name)", async function () {
    // A web component class cannot be registered multiple times.
    class CustomComponent extends HTMLElement {
        constructor() {
            super();
        }

        connectedCallback() {
            const greeting = this.getAttribute("greeting") ?? "";
            this.innerHTML = `<span class="hello">${greeting}</span>`;
        }
    }

    const tag = defineComponent(CustomComponent);
    const container1 = document.createElement("div");
    const container2 = document.createElement("div");
    document.body.append(container1, container2);

    const { node: node1 } = await renderComponent(tag, {
        attributes: {
            greeting: "Hello"
        },
        container: container1
    });

    const { node: node2 } = await renderComponent(tag, {
        attributes: {
            greeting: "Bye"
        },
        container: container2
    });

    expect([node1.innerHTML, node2.innerHTML]).toMatchSnapshot();
});

it("should render a web component with a shadow dom", async function () {
    class CustomComponent extends HTMLElement {
        constructor() {
            super();
            this.attachShadow({ mode: "open" }); // Must be open in tests for queries to work
        }

        connectedCallback() {
            this.shadowRoot!.innerHTML = `
                <div class="container">
                    <span class="hello">Hello World</span>
                </div>
            `;
        }
    }

    // Waits for the shadow dom to render and returns the inner container (and queries that search within in).
    const { node, innerContainer, queries } = await renderComponentShadowDOM(CustomComponent, {
        innerContainerSelector: ".container"
    });
    expect(node.shadowRoot).toBeDefined();
    expect(innerContainer.tagName).toEqual("DIV");
    expect(innerContainer.className).toEqual("container");

    // Searches in `.container`
    const span = await queries.findByText(/Hello/);
    expect(span).toMatchSnapshot();
});

it("should render a custom component into the dom", async () => {
    const elem = createCustomElement({
        component: () => createElement("div", { className: "test" }, "hello world")
    });

    // Waits until the component is rendered.
    // Returns an inner container from the shadow dom (`.pioneer-root` by default).
    // The queries object searches in that inner container.
    const { queries } = await renderComponentShadowDOM(elem);
    const div = await queries.findByText("hello world");
    expect(div.className).toBe("test");
});
