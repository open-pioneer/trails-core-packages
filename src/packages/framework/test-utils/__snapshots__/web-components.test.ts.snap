// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render a web component with a shadow dom 1`] = `
<span
  class="hello"
>
  Hello World
</span>
`;

exports[`should render the same component multiple times (needs shared tag name) 1`] = `
[
  "<span class=\\"hello\\">Hello</span>",
  "<span class=\\"hello\\">Bye</span>",
]
`;
